runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -5193,8 +5206,13 @@ TR::Register
    TR::LabelSymbol *fallThru   = generateLabelSymbol(cg);
    // Create the monitor exit snippet
    TR::LabelSymbol *snippetLabel = generateLabelSymbol(cg);
-   if (TR::Compiler->om.areValueTypesEnabled())
-       TR::TreeEvaluator::generateCheckForValueTypeMonitorEnterOrExit(node, snippetLabel, cg);
+
+   if (TR::Compiler->om.areValueTypesEnabled() && (cg->isMonitorValueType(node) == TR_maybe))
+      TR::TreeEvaluator::generateCheckForValueMonitorEnterOrExit(node, J9ClassIsValueType, snippetLabel, cg);
+
+   if (TR::Compiler->om.areValueBasedMonitorChecksEnabled() && (cg->isMonitorValueBased(node) == TR_maybe))
+      TR::TreeEvaluator::generateCheckForValueMonitorEnterOrExit(node, J9ClassIsValueBased, snippetLabel, cg);
```

Comment:
Just so I understand, the reason we can't combine these two into a single check is because the VM helper that will ultimately be called is only expecting one flag value or the other.  That is, it currently can't handle the situation when both flag values are set but it could be adapted to do so.

Reply:
> the reason we can't combine these two into a single check is because the VM helper that will ultimately be called is only expecting one flag value or the other. 

This is my understanding too. They're mutually exclusive. 

> That is, it currently can't handle the situation when both flag values are set but it could be adapted to do so.

The two flags shouldn't be set at the same time. But with the issue mentioned in https://github.com/eclipse/openj9/pull/11358#discussion_r538913598. both flags could be set with a build with`JAVA_SPEC_VERSION  >= 16` and `J9VM_OPT_VALHALLA_VALUE_TYPES` is defined. The VM helper will handle it as value type case regardless of which flag is set or what type of monitor object is (value based or value type). It's an issue. It should be prevented (https://github.com/eclipse/openj9/pull/11358#discussion_r539369193).

Reply:
> The two flags shouldn't be set at the same time. But with the issue mentioned in #11358 (comment). both flags could be set with a build with JAVA_SPEC_VERSION >= 16 and J9VM_OPT_VALHALLA_VALUE_TYPES is defined. The VM helper will handle it as value type case regardless of which flag is set or what type of monitor object is (value based or value type). It's an issue. It should be prevented (#11358 (comment)).

Just to clarify. At the moment I don't see us releasing a build where both `JAVA_SPEC_VERSION >= 16` and `J9VM_OPT_VALHALLA_VALUE_TYPES` are true since the purpose of JEP 390 is to aid migration to valuetypes. If this assumption is incorrect we will modify the helpers to address this. So from the JIT perspective, you do not need to worry about the dual cases. You can just test the mask (`J9_CLASS_DISALLOWS_LOCKING_FLAGS`) and call the helper. That is what the VM is currently doing. If supporting both modes simultaneously becomes a requirement, then we will have to change the VM to support this. And if you are testing the mask, it will be no effort on your part to support this.

Reply:
I'll look at how to combine the code in CG to test J9_CLASS_DISALLOWS_LOCKING_FLAGS instead of using two APIs